### PR TITLE
Global test timeout

### LINKED ISSUE #24066: Limit time tests can take when running in PR builds
I think we can safely set the limit to 5s.

We have recorded tests that take 30+ secs we need to catch them proactively.

/cc @JoshLove-msft

### PR DESCRIPTION
**Pull request by christothes**
closes #24066

### CODE REVIEW

File: eng/Directory.Build.Common.props [hunk 1]
```diff
@@ -243,6 +243,14 @@
     <RunSettingsFilePath Condition="'$(AZURE_SKIP_DEFAULT_RUN_SETTINGS)' != 'true'">$(RepoEngPath)\nunit.runsettings</RunSettingsFilePath>
   </PropertyGroup>
 
+  <!-- Set global test timeout of 5 seconds -->
+  <ItemGroup Condition="'$(IsTestProject)' == 'true'">
```
Comment (pakrym) [inline eng/Directory.Build.Common.props:247]: How do we override? Also what about live tests?

Comment (christothes) [reply eng/Directory.Build.Common.props:247]: The most local attribute takes precedence. So to override, a specific class, or test could be marked with a longer timeout.

For live tests, I'll look into if we can possibly set this dynamically based on test mode.

File: eng/Directory.Build.Common.props [hunk 2]
```diff
@@ -243,6 +243,14 @@
...
+    <AssemblyAttribute Include="NUnit.Framework.Timeout">
```
Comment (weshaggard) [inline eng/Directory.Build.Common.props:248]: Do we have test projects that are not using NUnit? If there are then this will likely fail to compile for them. I'm thinking about track 1 libraries for example.

Comment (christothes) [reply eng/Directory.Build.Common.props:248]: Yes, I made a change just now to eliminate track1 with the IsClientLibrary condition.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -4,4 +4,3 @@
 using NUnit.Framework;
 
 [assembly: Parallelizable(ParallelScope.All)]
-[assembly: NUnit.Framework.Timeout(5 * 60 * 1000)]
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs:7]: Your change greatly reduces this;  5 seconds is not long enough for the Live tests and probably too short for some of the background operations for the Processor as well.

File: eng/Directory.Build.Common.props [hunk 4]
```diff
@@ -243,6 +243,14 @@
     <RunSettingsFilePath Condition="'$(AZURE_SKIP_DEFAULT_RUN_SETTINGS)' != 'true'">$(RepoEngPath)\nunit.runsettings</RunSettingsFilePath>
   </PropertyGroup>
 
+  <!-- Set global test timeout of 5 seconds -->
+  <ItemGroup Condition="'$(IsTestProject)' == 'true' and '$(IsClientLibrary)' == 'true'">
```
Comment (jsquire) [inline eng/Directory.Build.Common.props:247]: 5 seconds is really short, won't this break a ton of live tests?

Comment (jsquire) [reply eng/Directory.Build.Common.props:247]: We'll need an opt-out for this in Event Hubs and Service Bus, at least.

**Review by jsquire — CHANGES_REQUESTED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (christothes) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs:7]: Live tests will be addressed as part of this PR, or I'll abandon this approach. 

For the other tests, we can just add a more local attribute on those specific tests.

File: eng/Directory.Build.Common.props (hunk 4 above)
Comment (christothes) [reply eng/Directory.Build.Common.props:247]: see above comment 😄

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs:7]: I'd prefer to opt-out at the service directory level for Event Hubs.   I'll let Josh make the call for Service Bus.

Comment (christothes) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs:7]: The intent of this change is to avoid long running unit tests. Wouldn't opt in be better (it should be an exception and very intentional to check in a long running test)

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs [hunk 5]
```diff
@@ -39,6 +39,12 @@ public FallbackCommand(TestCommand innerCommand) : base(innerCommand)
             { }
             public override TestResult Execute(TestExecutionContext context)
             {
+                // Lift the global timeout setting if we are not in Playback mode
+                var recordedTest = context.TestObject as RecordedTestBase;
+                if (recordedTest?.Mode != RecordedTestMode.Playback)
+                {
+                    context.TestCaseTimeout = 0; // set the default timeout;
```
Comment (christothes) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs:46]: experimental at this point. If this works well, we may want to walk up the class hierarchy to find the lowest level Timeout attribute below assembly level set and use that, if one exists.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs:7]: As long as this timeout doesn't apply to live tests, I think it should be okay.

File: eng/Directory.Build.Common.props (hunk 2 above)
Comment (weshaggard) [reply eng/Directory.Build.Common.props:248]: I would use the IsClientLIbrary variable directly instead of trying to duplicate the condition hear.

Comment (christothes) [reply eng/Directory.Build.Common.props:248]: I did initially, but I think some track1 clients set this property directly and the name check eliminates those.

Comment (weshaggard) [reply eng/Directory.Build.Common.props:248]: If there are libraries setting that directly then they should not be track 1 and so they should be considered track 2 and we should probably treat them the same way.

Comment (heaths) [reply eng/Directory.Build.Common.props:249]: Did you take into account that we have many track 2s that use "Microsoft.Azure." as a prefix, though? It would be hard to eliminate them all. E.g. sdk/core/Microsoft.Azure.Core.Spatial.

Comment (Mohit-Chakraborty) [reply eng/Directory.Build.Common.props:249]: Agree. The evaluation condition of $(IsClientLibrary) is unfortunately not entirely correct. So, we are forced to set it to true in some track 2 libraries.
So, the current best option is to continue using this one variable, so we don't accidentally exclude any track 2 libraries.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeyVaultOnlyAttribute.cs [hunk 6]
```diff
@@ -23,10 +23,14 @@ public TestCommand Wrap(TestCommand command)
             {
                 t.RunState = RunState.Skipped;
                 test.Properties.Set(PropertyNames.SkipReason, $"This test can only run on Key Vault.");
+                // Timeout cannot be set on SkipCommands
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeyVaultOnlyAttribute.cs:26]: Why is this even necessary, though? The whole test would be skipped, so maybe just don't set the timeout at all?

**Review by heaths — CHANGES_REQUESTED** [review]

File: sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeyVaultOnlyAttribute.cs (hunk 6 above)
Comment (christothes) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeyVaultOnlyAttribute.cs:26]: Seems to be an issue with how NUnit implements the TimeoutCommand. It throws without this change here: 
https://github.com/nunit/nunit/blob/f0e35aa13e2e87fe9ad641115385afccab4bb5b5/src/NUnitFramework/framework/Internal/Commands/TimeoutCommand.cs#L38

File: eng/Directory.Build.Common.props (hunk 2 above)
Comment (christothes) [reply eng/Directory.Build.Common.props:249]: I've implemented it as opt out now using the new `IsOptOutOfGlobalTestTimeout` flag.

**Review by heaths — APPROVED** [review]

File: sdk/eventhub/Directory.Build.props [hunk 7]
```diff
@@ -10,6 +10,7 @@
   <PropertyGroup Condition="$(MSBuildProjectName.StartsWith('Azure.')) and $(MSBuildProjectName.Contains('.Samples'))">
     <IsSamplesProject>true</IsSamplesProject>
     <IsTestSupportProject>true</IsTestSupportProject>
+    <IsOptOutOfGlobalTestTimeout>true</IsOptOutOfGlobalTestTimeout>
   </PropertyGroup>
```
Comment (jsquire) [inline sdk/eventhub/Directory.Build.props:14]:
```suggestion
  <PropertyGroup Condition="$(MSBuildProjectName.StartsWith('Azure.')) and $(MSBuildProjectName.Contains('.Samples'))">
    <IsSamplesProject>true</IsSamplesProject>
    <IsTestSupportProject>true</IsTestSupportProject>
  </PropertyGroup>
  
  <!-- 
    OptOut of the global NUnit test timeout for all Event Hubs libraries.
    
    The approach that NUnit uses to measure timing for tests does not play well with parallel execution
    and highly asynchronous code.  Many of the Event Hubs tests run against non-deterministic background
    operations where timing is longer than normal unit tests and the exact timing is difficult to pin down.
    
    Tests currently use an environment variable with cancellation-based timeout control, which allows for
    more generous timeouts in the CI environments, where resources are often more constrained and 
    continuations take longer to be scheduled.    
  -->
  <PropertyGroup>
        <IsOptOutOfGlobalTestTimeout>true</IsOptOutOfGlobalTestTimeout>
  </PropertyGroup>
```

We probably want to slide this out into its own group;  the conditionals on this one specifically target sample projects.

**Review by jsquire — APPROVED** [review]
Event Hubs bits look good to me.    

We probably want to run the Service Bus live tests to ensure that they play nice with the new value, since they also need to bootstrap Azure resources as part of each test.

File: sdk/core/Azure.Core/tests/Base64UrlTests.cs [hunk 8]
```diff
@@ -60,7 +60,11 @@ public void ValidateEncodeDecodeRandom()
 
                     var decoded = Base64Url.Decode(encoded);
 
-                    CollectionAssert.AreEqual(data, decoded, "Data round trip failed. Seed {0}", seed);
+                    Assert.AreEqual(data.Length, decoded.Length, $"Data round trip failed. Seed {seed}");
+                    for (int k = 0; k < data.Length; k++)
+                    {
+                        Assert.True(data[k] == decoded[k], $"Data round trip failed. Seed {seed}");
```
Comment (christothes) [inline sdk/core/Azure.Core/tests/Base64UrlTests.cs:66]: believe it or not, this change saves 2 seconds of execution time.

File: sdk/core/Azure.Core/tests/TransportFunctionalTests.cs [hunk 9]
```diff
@@ -94,7 +94,7 @@ public async Task SettingHeaderOverridesDefaultContentLength()
         public async Task CanSetContentLengthOverMaxInt()
         {
             long contentLength = 0;
-            using TestServer testServer = new TestServer(
+            TestServer testServer = new TestServer(
```
Comment (christothes) [inline sdk/core/Azure.Core/tests/TransportFunctionalTests.cs:97]: disposing the TestServer for these two tests adds 4 seconds to the test. Gave up trying to figure out why.

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs [hunk 10]
```diff
@@ -191,7 +191,7 @@ public async Task ContainerRegistryChallengeAuthenticationPolicy_RefreshesRefres
             int refreshTokenRequests = 0;
             int accessTokenRequests = 0;
             TimeSpan refreshOffset = TimeSpan.FromSeconds(30);
-            TimeSpan expiryTime = TimeSpan.FromSeconds(35);
+            TimeSpan expiryTime = TimeSpan.FromSeconds(32);
```
Comment (annelo-msft) [inline sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs:194]: Just curious: why this change?

File: sdk/eventhub/Directory.Build.props [hunk 11]
```diff
@@ -10,7 +10,21 @@
   <PropertyGroup Condition="$(MSBuildProjectName.StartsWith('Azure.')) and $(MSBuildProjectName.Contains('.Samples'))">
     <IsSamplesProject>true</IsSamplesProject>
     <IsTestSupportProject>true</IsTestSupportProject>
-    <IsOptOutOfGlobalTestTimeout>true</IsOptOutOfGlobalTestTimeout>
+  </PropertyGroup>
+  
+  <!-- 
+    OptOut of the global NUnit test timeout for all Event Hubs libraries.
```
Comment (heaths) [inline sdk/eventhub/Directory.Build.props:16]: Nit: "Opt out" would be the written verb form.

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs (hunk 10 above)
Comment (christothes) [reply sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs:194]: It eliminates a few seconds of time from the test, which gets it under the new global timeout of 5 seconds.

**Comment by christothes** [comment]
> Event Hubs bits look good to me.
> 
> We probably want to run the Service Bus live tests to ensure that they play nice with the new value, since they also need to bootstrap Azure resources as part of each test.

Thanks - I plan to run everything live before letting this in.

File: sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs [hunk 12]
```diff
@@ -121,6 +121,12 @@ public Response GetResponse(RecordEntry recordEntry)
                 {
                     response.AddHeader(new HttpHeader(responseHeader.Key, value));
                 }
+
+                // Ignore the Retry-After header wait time for 429 responses during playback.
+                if (response.Status == 429)
+                {
+                    response.AddHeader(new HttpHeader("Retry-After", "0"));
+                }
```
Comment (christothes) [inline sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs:129]: This works nicely, but seems wrong. Thoughts?

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs:129]: Why are we adding this?

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs:129]: Ah nevermind - nice comment..

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs:129]: Should we be adding the header or updating it?

Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs:129]: Updating would be ideal.

Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs]: Changed to  update.

File: sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs [hunk 13]
```diff
@@ -121,6 +121,12 @@ public Response GetResponse(RecordEntry recordEntry)
                 {
                     response.AddHeader(new HttpHeader(responseHeader.Key, value));
                 }
+
+                // Ignore the Retry-After header wait time for 429 responses during playback.
+                if (response.Status == 429 && response.Headers.TryGetValue("Retry-After", out _))
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs:126]: This won't work when we switch to an out-of-process test proxy. We might need to file a feature request for them.

Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/PlaybackTransport.cs:126]: cc @JoshLove-msft

File: eng/Directory.Build.Common.props [hunk 14]
```diff
@@ -243,6 +244,14 @@
     <RunSettingsFilePath Condition="'$(AZURE_SKIP_DEFAULT_RUN_SETTINGS)' != 'true'">$(RepoEngPath)\nunit.runsettings</RunSettingsFilePath>
   </PropertyGroup>
 
+  <!-- Set global test timeout of 5 seconds -->
```
Comment (pakrym) [inline eng/Directory.Build.Common.props:247]:
```suggestion
  <!-- Set global test timeout of 10 seconds -->
```

File: sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs (hunk 10 above)
Comment (pakrym) [reply sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs:194]: Is this going to be flaky? Are we barely under the timeout?

Comment (christothes) [reply sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs:194]: It was never flaky in the many times it was run , but since then the timeout has doubled so the risk should be low.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/DocumentAnalysisClient/DocumentAnalysisClientTests.cs [hunk 15]
```diff
@@ -36,7 +36,7 @@ private DocumentAnalysisClient CreateClient()
             var fakeEndpoint = new Uri("http://notreal.azure.com");
             var fakeCredential = new AzureKeyCredential("fakeKey");
 
-            return new DocumentAnalysisClient(fakeEndpoint, fakeCredential);
+            return new DocumentAnalysisClient(fakeEndpoint, fakeCredential, new DocumentAnalysisClientOptions(){ Retry = { Delay = TimeSpan.Zero, Mode = RetryMode.Fixed}} );
```
Comment (pakrym) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/tests/DocumentAnalysisClient/DocumentAnalysisClientTests.cs:39]: InstrumentClientOptions?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/E2eTestBase.cs [hunk 16]
```diff
@@ -38,7 +39,7 @@ protected DigitalTwinsClient GetClient(DigitalTwinsClientOptions options = null)
         {
             if (options == null)
             {
-                options = new DigitalTwinsClientOptions();
+                options = new DigitalTwinsClientOptions(){ Retry = { Delay = TimeSpan.Zero, Mode = RetryMode.Fixed}};
```
Comment (pakrym) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/E2eTestBase.cs:42]: InstrumentClientOptions?

File: sdk/identity/Azure.Identity/tests/AsyncLockWithValueTests.cs [hunk 17]
```diff
@@ -131,7 +132,7 @@ public async Task AsyncLockWithValue_GetLockOrValueAsync_OneHundredCalls_HasNoVa
             var alwv = new AsyncLockWithValue<int>();
             var firstLock = await alwv.GetLockOrValueAsync(async);
             var tasks = new List<Task>();
-            for (var i = 0; i < 100; i++)
+            for (var i = 0; i < 50; i++)
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/AsyncLockWithValueTests.cs:135]: Do we need both the attribute and the values change?

Comment (christothes) [reply sdk/identity/Azure.Identity/tests/AsyncLockWithValueTests.cs:135]: probably not, but the choice of 100 seems fairly arbitrary and increases the test duration.

File: sdk/iot/Azure.IoT.Hub.Service/tests/QueryTests.cs [hunk 18]
```diff
@@ -127,8 +131,11 @@ await foreach (TwinData item in queryResponse)
 
                     tryCount++;
 
-                    // Adding a delay to account for query cache sync.
-                    await Task.Delay(TimeSpan.FromSeconds(5)).ConfigureAwait(false);
+                    if (Mode != RecordedTestMode.Playback)
+                    {
+                        // Adding a delay to account for query cache sync.
+                        await Task.Delay(TimeSpan.FromSeconds(5)).ConfigureAwait(false);
```
Comment (pakrym) [inline sdk/iot/Azure.IoT.Hub.Service/tests/QueryTests.cs:137]: Do we need a helper for this in the TestFrameowrk? RecordedTestBase.Delay?

Comment (christothes) [reply sdk/iot/Azure.IoT.Hub.Service/tests/QueryTests.cs:137]: @heaths Suggested the same thing. I think it's a good idea, but seemed rare enough in this PR not to be a big deal. Perhaps I'll do it.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs [hunk 19]
```diff
@@ -9,11 +9,11 @@
 
 namespace Azure.Core.TestFramework
 {
-    [AttributeUsage(AttributeTargets.Method)]
     /// <summary>
     /// This attribute replaces the [Test] attribute and will dynamically re-record recorded tests on failure.
     /// Tests that are re-recorded will complete with a error status and indicate that copying the updated recording to SessionRecords is needed.
     /// </summary>
+    [AttributeUsage(AttributeTargets.Method)]
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs:16]: What if a test doesn't use the `RecordedTestAttribute ` ?

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialTests.cs [hunk 20]
```diff
@@ -376,7 +376,7 @@ public async Task VerifyInitialImdsConnectionTimeoutHonored()
         {
             using var server = new TestServer(async context =>
             {
-                await Task.Delay(8000);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialTests.cs:379]: Was 8000 tied to some default timeout in MSI endpoint discovery?

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Listeners/ScanBlobScanLogHybridPollingStrategyTests.cs [hunk 21]
```diff
@@ -25,6 +25,7 @@
 
 namespace Microsoft.Azure.WebJobs.Extensions.Storage.Blobs.Listeners
 {
+    [NonParallelizable]
```
Comment (pakrym) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Listeners/ScanBlobScanLogHybridPollingStrategyTests.cs:28]: Why?

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs (hunk 19 above)
Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs:16]: Then it will get the global timeout.

File: sdk/core/Azure.Core.TestFramework/src/TestServer.cs [hunk 22]
```diff
@@ -45,8 +45,8 @@ public TestServer(RequestDelegate app, bool https = false)
                     services.AddSingleton<IStartup>(this);
                 })
                 .UseSetting(WebHostDefaults.ApplicationKey, typeof(TestServer).GetTypeInfo().Assembly.FullName)
+                .UseShutdownTimeout(TimeSpan.FromMilliseconds(1000))
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/TestServer.cs:48]: I may have fixed both tests that required this.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Listeners/ScanBlobScanLogHybridPollingStrategyTests.cs (hunk 21 above)
Comment (christothes) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Listeners/ScanBlobScanLogHybridPollingStrategyTests.cs:28]: I should remove. This was an experiment to see if it helped with the thread contention which was causing so much variability with tests with high thread contention.

File: sdk/core/Azure.Core.TestFramework/src/TestServer.cs (hunk 22 above)
Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/TestServer.cs:48]: Is there any down side to keeping it?

File: sdk/eventhub/Directory.Build.props [hunk 23]
```diff
@@ -11,6 +11,21 @@
...
+    The approach that NUnit uses to measure timing for tests does not play well with parallel execution
+    and highly asynchronous code.  Many of the Event Hubs tests run against non-deterministic background
+    operations where timing is longer than normal unit tests and the exact timing is difficult to pin down.
+    
+    Tests currently use an environment variable with cancellation-based timeout control, which allows for
+    more generous timeouts in the CI environments, where resources are often more constrained and 
+    continuations take longer to be scheduled.    
+  -->
+  <PropertyGroup>
+        <IsOptOutOfGlobalTestTimeout>true</IsOptOutOfGlobalTestTimeout>
```
Comment (pakrym) [inline sdk/eventhub/Directory.Build.props:27]: Should we allow increasing timeouts instead of complete opt-out?

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialTests.cs (hunk 20 above)
Comment (christothes) [reply sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialTests.cs:379]: @schaabs Do you recall why 8 seconds was chosen?

File: sdk/eventhub/Directory.Build.props (hunk 23 above)
Comment (christothes) [reply sdk/eventhub/Directory.Build.props:27]: I think OptOut should be considered test debt that needs to be fixed at some point. I only used this in cases where the work to address the timeout issues was too much for this PR. Eventhubs is the best example of this, because of how different its use of test infrastructure is.

File: sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs (hunk 19 above)
Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/RecordedTestAttribute.cs:16]: Even for a live run? Am I correct that this makes the [RecordedTest] attribute mandatory?

File: sdk/core/Azure.Core.TestFramework/src/ClientTestBase.cs [hunk 24]
```diff
@@ -27,6 +31,18 @@ public ClientTestBase(bool isAsync)
             IsAsync = isAsync;
         }
 
+        [TearDown]
+        public virtual void GlobalTimeoutTearDown()
+        {
+            var context = TestContext.CurrentContext;
+            var executionContext = executionContextField.GetValue(context) as TestExecutionContext;
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/ClientTestBase.cs:38]: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/TestExecutionContext.cs#L129 ?

File: sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs [hunk 25]
```diff
@@ -14,6 +14,7 @@
 
 namespace Azure.Core.Tests
 {
+    [Timeout(30000)]
```
Comment (pakrym) [inline sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs:17]: rem?

File: sdk/core/Azure.Core.TestFramework/src/TestServer.cs (hunk 22 above)
Comment (pakrym) [reply sdk/core/Azure.Core.TestFramework/src/TestServer.cs]: I've discovered a few response lifetime bugs because kestrel took a long time to shut down signaling that not all connections were closed.

File: sdk/iot/Azure.IoT.Hub.Service/tests/QueryTests.cs (hunk 18 above)
Comment (pakrym) [reply sdk/iot/Azure.IoT.Hub.Service/tests/QueryTests.cs:137]: It seems to be the most common pattern to slow tests down.

File: sdk/core/Azure.Core.TestFramework/src/ClientTestBase.cs (hunk 24 above)
Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/ClientTestBase.cs:38]: nice find!

File: sdk/iot/Azure.IoT.Hub.Service/tests/QueryTests.cs (hunk 18 above)
Comment (christothes) [reply sdk/iot/Azure.IoT.Hub.Service/tests/QueryTests.cs:137]: I've implemented it, just forgot to use it here.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (pakrym) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Properties/AssemblyInfo.cs:7]: revert?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/E2eTestBase.cs (hunk 16 above)
Comment (christothes) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/E2eTestBase.cs:42]: it's actually happening a few lines below

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/DocumentAnalysisClient/DocumentAnalysisClientTests.cs (hunk 15 above)
Comment (christothes) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/tests/DocumentAnalysisClient/DocumentAnalysisClientTests.cs:39]: This is not in RecordedTestBase.
